### PR TITLE
Fallback to system headers for future CUDA 12.x versions

### PR DESCRIPTION
**Pull request by kmaehashi**
Do not use bundled headers for unsupported future CUDA 12.x releases, as headers from the bundled `cuda-12` directory are unlikely to work with them.

Once new CUDA version is out, users can either install the full CTK with headers or do `pip install -U nvidia-cuda-runtime-cu12` to continue using CuPy without waiting for a new release.

cc/ @leofang

### CODE REVIEW

**Comment by jakirkham** [comment]
Thanks Kenichi! 🙏

Is this being backported as well?

**Comment by kmaehashi** [comment]
Yes, I'd backport this one to v13.

File: cupy/_core/core.pyx [hunk 1]
```diff
@@ -2215,24 +2217,34 @@ cpdef tuple assemble_cupy_compiler_options(tuple options):
...
+                    # Unsupported CUDA 12.x variant
+                    _bundled_include = None
             else:
                 # CUDA versions not yet supported.
                 _bundled_include = None
 
-        if _bundled_include is None and _cuda_path is None:
+            # Check if headers from cudart wheels are available.
+            wheel_dir_count = len(
+                _environment._get_include_dir_from_wheel(major, minor))
```
Comment (jakirkham) [inline cupy/_core/core.pyx:2241]: Think this now needs to be

```suggestion
                _environment._get_include_dir_from_conda_or_wheel(major, minor))
```

**Review by jakirkham — COMMENTED** [review]
Thanks Kenichi! 🙏

Think the function name needs a tweak after PR: https://github.com/cupy/cupy/pull/8519

Included that suggestion below

File: cupy/_core/core.pyx [hunk 2]
```diff
@@ -2215,24 +2217,34 @@ cpdef tuple assemble_cupy_compiler_options(tuple options):
...
+                    # Unsupported CUDA 12.x variant
+                    _bundled_include = None
             else:
                 # CUDA versions not yet supported.
                 _bundled_include = None
 
-        if _bundled_include is None and _cuda_path is None:
+            # Check if headers from cudart wheels are available.
+            wheel_dir_count = len(
+                _environment._get_include_dir_from_conda_or_wheel(major, minor))
```
Comment (jakirkham) [inline cupy/_core/core.pyx:2241]: [The style check]( https://github.com/cupy/cupy/actions/runs/10465211580/job/28980078200?pr=8518#step:6:24 ) says this is 1 character over

```
cupy/_core/core.pyx:2241:80: E501 line too long (80 > 79 characters)
```

So let's wrap one

```suggestion
            wheel_dir_count = len(
                _environment._get_include_dir_from_conda_or_wheel(major, minor)
            )
```

File: cupy/_core/core.pyx [hunk 3]
```diff
@@ -2215,24 +2217,35 @@ cpdef tuple assemble_cupy_compiler_options(tuple options):
...
-            elif _cuda_major == 12:
-                _, minor = nvrtc.getVersion()
+            elif major == 12:
                 # TODO(leofang): update the upper bound when a new release
                 # is out
-                if minor < 2 or minor > 6:
+                if minor < 2:
                     _bundled_include = 'cuda-12'
-                else:
+                elif minor < 6:
```
Comment (kmaehashi) [inline cupy/_core/core.pyx:2230]:
```suggestion
                elif minor < 7:
```

**Pull request merged**